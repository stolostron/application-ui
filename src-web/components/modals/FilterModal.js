/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import { Modal } from 'carbon-components-react'
import NestedTable from '../common/NestedTable'
import resources from '../../../lib/shared/resources'
import msgs from '../../../nls/platform.properties'

resources(() => {
  require('../../../scss/modal.scss')
})

class FilterModal extends React.PureComponent {
  constructor(props) {
    super(props)
    this.handleSubmitClick = this.handleSubmitClick.bind(this)
    this.selectionChanged = this.selectionChanged.bind(this)
    this.state = {
      tags: props.selected || []
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps && nextProps.selected !== this.props.selected) {
      this.setState({ tags: nextProps.selected })
    }
  }

  convertFilterArray(queries) {
    const filterArray = []
    for (const [type, value] of Object.entries(queries)) {
      filterArray.push({
        key: type,
        value
      })
    }
    return filterArray
  }

  createFilterTable(availableFilters, selectedFilters, selectionChanged) {
    return (
      <NestedTable
        availableItems={this.convertFilterArray(availableFilters)}
        header={msgs.get('modal.formfield.name', this.context.locale)}
        selectedItems={selectedFilters}
        selectionChanged={selectionChanged}
      />
    )
  }

  selectionChanged(selections) {
    this.setState({ tags: selections })
  }

  handleSubmitClick() {
    const { handleModalSubmit } = this.props
    handleModalSubmit(this.state.tags)
  }

  render() {
    const { availableFilters = [], handleModalClose } = this.props
    const { modalOpen } = this.props
    return (
      <div>
        <Modal
          className="modal-with-editor"
          open={modalOpen}
          modalHeading={msgs.get(
            'modal.resource-filter.label',
            this.context.locale
          )}
          primaryButtonText={msgs.get('actions.save', this.context.locale)}
          primaryButtonDisabled={false}
          secondaryButtonText={msgs.get('actions.cancel', this.context.locale)}
          onRequestSubmit={this.handleSubmitClick}
          onRequestClose={handleModalClose}
        >
          {this.createFilterTable(
            availableFilters,
            this.state.tags,
            this.selectionChanged
          )}
        </Modal>
      </div>
    )
  }
}

FilterModal.propTypes = {
  availableFilters: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  handleModalClose: PropTypes.func,
  handleModalSubmit: PropTypes.func,
  modalOpen: PropTypes.bool,
  selected: PropTypes.array
}

export default FilterModal
