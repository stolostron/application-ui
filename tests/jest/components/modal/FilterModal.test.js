/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
"use strict";

import React from "react";
import { shallow } from "enzyme";
import FilterModal from "../../../../src-web/components/modals/FilterModal";
import { availableFilters } from "./ModalsTestingData";

describe("FilterModal modal", () => {
  const handleModalClose = jest.fn();
  const handleModalSubmit = jest.fn();
  it("renders as expected", () => {
    const component = shallow(
      <FilterModal handleModalClose={handleModalClose} />
    );
    expect(component).toMatchSnapshot();
  });
  it("componentWillReceiveProps as expected", () => {
    const component = shallow(
      <FilterModal
        availableFilters={availableFilters}
        handleModalClose={handleModalClose}
      />
    );
    component.instance().UNSAFE_componentWillReceiveProps();
    expect(component.state().tags).toEqual([]);
    component
      .instance()
      .UNSAFE_componentWillReceiveProps({ selected: ["test"] });
    expect(component.state().tags).toEqual(["test"]);
  });
  it("convertFilterArray as expected", () => {
    const component = shallow(
      <FilterModal
        availableFilters={availableFilters}
        handleModalClose={handleModalClose}
      />
    );
    expect(
      component.instance().convertFilterArray(availableFilters)
    ).toMatchSnapshot();
  });
  it("convertFilterArray as expected", () => {
    const component = shallow(
      <FilterModal
        availableFilters={availableFilters}
        handleModalClose={handleModalClose}
      />
    );
    expect(
      component
        .instance()
        .createFilterTable(
          availableFilters,
          ["testConvertFilterArray"],
          jest.fn()
        )
    ).toMatchSnapshot();
  });
  it("selectionChanged as expected", () => {
    const component = shallow(
      <FilterModal
        availableFilters={availableFilters}
        handleModalClose={handleModalClose}
      />
    );
    component
      .instance()
      .selectionChanged(["testSelectionChanged1", "testSelectionChanged2"]);
    expect(component.state()).toMatchSnapshot();
  });
  it("handleSubmitClick as expected", () => {
    const component = shallow(
      <FilterModal
        availableFilters={availableFilters}
        handleModalClose={handleModalClose}
        handleModalSubmit={handleModalSubmit}
      />
    );
    component.instance().handleSubmitClick();
    expect(component.state()).toMatchSnapshot();
  });
});
