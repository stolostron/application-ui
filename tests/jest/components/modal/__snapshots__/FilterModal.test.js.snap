// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateResourceModal modal convertFilterArray as expected 1`] = `
Array [
  Object {
    "key": "standards",
    "value": Object {
      "availableSet": Array [
        "FISMA",
        "PCI",
        "HIPAA",
        "NIST",
      ],
      "name": "Standards",
    },
  },
  Object {
    "key": "categories",
    "value": Object {
      "availableSet": Array [
        "System And Communications Protections",
        "System And Information Integrity",
      ],
      "name": "Categories",
    },
  },
  Object {
    "key": "controls",
    "value": Object {
      "availableSet": Array [
        "Secret Encryption",
        "Cert Manager",
        "Mutation Advisor",
        "VA",
      ],
      "name": "Controls",
    },
  },
  Object {
    "key": "type",
    "value": Object {
      "availableSet": Array [
        "Enforce",
        "Inform only",
      ],
      "name": "Type",
    },
  },
  Object {
    "key": "severity",
    "value": Object {
      "availableSet": Array [],
    },
  },
]
`;

exports[`CreateResourceModal modal convertFilterArray as expected 2`] = `
<NestedTable
  availableItems={
    Array [
      Object {
        "key": "standards",
        "value": Object {
          "availableSet": Array [
            "FISMA",
            "PCI",
            "HIPAA",
            "NIST",
          ],
          "name": "Standards",
        },
      },
      Object {
        "key": "categories",
        "value": Object {
          "availableSet": Array [
            "System And Communications Protections",
            "System And Information Integrity",
          ],
          "name": "Categories",
        },
      },
      Object {
        "key": "controls",
        "value": Object {
          "availableSet": Array [
            "Secret Encryption",
            "Cert Manager",
            "Mutation Advisor",
            "VA",
          ],
          "name": "Controls",
        },
      },
      Object {
        "key": "type",
        "value": Object {
          "availableSet": Array [
            "Enforce",
            "Inform only",
          ],
          "name": "Type",
        },
      },
      Object {
        "key": "severity",
        "value": Object {
          "availableSet": Array [],
        },
      },
    ]
  }
  header="Name"
  selectedItems={
    Array [
      "testConvertFilterArray",
    ]
  }
  selectionChanged={[MockFunction]}
/>
`;

exports[`CreateResourceModal modal handleSubmitClick as expected 1`] = `
Object {
  "tags": Array [],
}
`;

exports[`CreateResourceModal modal renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FilterModal
    handleModalClose={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Modal
        className="modal-with-editor"
        focusTrap={true}
        iconDescription="close the modal"
        modalHeading="Resource Filter"
        modalLabel=""
        onKeyDown={[Function]}
        onRequestClose={[MockFunction]}
        onRequestSubmit={[Function]}
        passiveModal={false}
        primaryButtonDisabled={false}
        primaryButtonText="Save"
        secondaryButtonText="Cancel"
        selectorPrimaryFocus="[data-modal-primary-focus]"
      >
        <NestedTable
          availableItems={Array []}
          header="Name"
          selectedItems={Array []}
          selectionChanged={[Function]}
        />
      </Modal>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <NestedTable
          availableItems={Array []}
          header="Name"
          selectedItems={Array []}
          selectionChanged={[Function]}
        />,
        "className": "modal-with-editor",
        "focusTrap": true,
        "iconDescription": "close the modal",
        "modalHeading": "Resource Filter",
        "modalLabel": "",
        "onKeyDown": [Function],
        "onRequestClose": [MockFunction],
        "onRequestSubmit": [Function],
        "open": undefined,
        "passiveModal": false,
        "primaryButtonDisabled": false,
        "primaryButtonText": "Save",
        "secondaryButtonText": "Cancel",
        "selectorPrimaryFocus": "[data-modal-primary-focus]",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "availableItems": Array [],
          "header": "Name",
          "selectedItems": Array [],
          "selectionChanged": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Modal
          className="modal-with-editor"
          focusTrap={true}
          iconDescription="close the modal"
          modalHeading="Resource Filter"
          modalLabel=""
          onKeyDown={[Function]}
          onRequestClose={[MockFunction]}
          onRequestSubmit={[Function]}
          passiveModal={false}
          primaryButtonDisabled={false}
          primaryButtonText="Save"
          secondaryButtonText="Cancel"
          selectorPrimaryFocus="[data-modal-primary-focus]"
        >
          <NestedTable
            availableItems={Array []}
            header="Name"
            selectedItems={Array []}
            selectionChanged={[Function]}
          />
        </Modal>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <NestedTable
            availableItems={Array []}
            header="Name"
            selectedItems={Array []}
            selectionChanged={[Function]}
          />,
          "className": "modal-with-editor",
          "focusTrap": true,
          "iconDescription": "close the modal",
          "modalHeading": "Resource Filter",
          "modalLabel": "",
          "onKeyDown": [Function],
          "onRequestClose": [MockFunction],
          "onRequestSubmit": [Function],
          "open": undefined,
          "passiveModal": false,
          "primaryButtonDisabled": false,
          "primaryButtonText": "Save",
          "secondaryButtonText": "Cancel",
          "selectorPrimaryFocus": "[data-modal-primary-focus]",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "availableItems": Array [],
            "header": "Name",
            "selectedItems": Array [],
            "selectionChanged": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": false,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": true,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`CreateResourceModal modal selectionChanged as expected 1`] = `
Object {
  "tags": Array [
    "testSelectionChanged1",
    "testSelectionChanged2",
  ],
}
`;
